File: csharp/bank-account/BankAccount.cs
public class BankAccount
{
    Mutex mutex = new Mutex(); // Provides atomic access.
    bool isOpen = false;    // Is the account open.
    decimal balance;        // The balance of the account.

    /// <summary>
    /// Open an account.
    /// </summary>
    /// <exception cref="InvalidOperationException">Fails if the account is already open.</exception>
    public void Open()
    {
        if (isOpen)
            throw new InvalidOperationException("Account is already open.");
        isOpen = true;
        balance = 0;
    }

    /// <summary>
    /// Close this account.
    /// </summary>
    /// <exception cref="InvalidOperationException">Fails is the account is not open.</exception>
    public void Close()
    {
        if (!isOpen)
            throw new InvalidOperationException("Cannot close an Account that is not open.");
        isOpen = false;
    }

    /// <summary>
    /// Return the account balance.
    /// </summary>
    public decimal Balance
    {
        get
        {
            if (isOpen)
            {
                try
                {
                    mutex.WaitOne();
                    return balance;
                }
                finally
                {
                    mutex.ReleaseMutex();
                }
            }
            else
                throw new InvalidOperationException("Account is not open.");
        }
    }

    /// <summary>
    /// Add money to the account.
    /// </summary>
    /// <param name="change">The amount to add.</param>
    /// <exception cref="InvalidOperationException">Fails if amount is less than or equal to zero.</exception>
    public void Deposit(decimal change)
    {
        if (change <= 0)
            throw new InvalidOperationException("Change must be a positive non-zero amount.");

        if (isOpen)
        {
            try
            {
                mutex.WaitOne();
                balance += change;
            }
            finally
            {
                mutex.ReleaseMutex();
            }
        }
        else
            throw new InvalidOperationException("Account is not open.");

    }

    /// <summary>
    /// Remove money from the account.
    /// </summary>
    /// <param name="change">The amount to remove.</param>
    /// <exception cref="InvalidOperationException">Fails if the change is less than or equal to zero. Also will fail if the amount to withdraw exceeds the balance.</exception>
    public void Withdraw(decimal change)
    {
        if (isOpen)
        {
            if (balance - change < 0)
                throw new InvalidOperationException("Cannot withdraw in excess of balance.");
            else if (change <= 0)
                throw new InvalidOperationException("Change must be a positive non-zero amount.");
            try
            {
                mutex.WaitOne();
                balance -= change;
            }
            finally
            {
                mutex.ReleaseMutex();
            }
        }
        else
            throw new InvalidOperationException("Account is not open.");
    }
}

File: go/bank-account/bank_account.go
// Bank Account example. Used to test concurrency with mutexes.
package bankaccount

import "sync"

// Account data
type Account struct {
	balance int64
	opened  bool
	mu      sync.Mutex
}

// Open an account and return a reference to it.
// @param amt: Initial account deposit amount.
// @returns: nil if initial deposit is negative otherwise a reference to the new account.
func Open(amt int64) *Account {
	if amt < 0 {
		return nil
	}
	return &Account{balance: amt, opened: true}
}

// Return the account balance.
// @returns: Balance of the account and success flag.
// @raises: Unsuccessful if acount not opened.
func (a *Account) Balance() (bal int64, ok bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.opened {
		return 0, false
	}
	return a.balance, true
}

// Deposit funds into the account. Calls Withdraw if amount is less than or equal to zero
// @param amt: The amount to deposit.
// @returns: The updated balance and a success flag.
// @raises: Unsuccessful if account not opened or any withdrawl problem if amount <= 0.
func (a *Account) Deposit(amt int64) (newBal int64, ok bool) {
	if amt <= 0 {
		return a.Withdraw(amt * -1)
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.opened {
		return 0, false
	}
	a.balance += amt
	return a.balance, true
}

// Withdraw amount from account balance
// @param amt: The amount to withdraw
// @returns: Remainining balance and success flag
// @raises: Not successful if insufficient funds, negative withdrawl, or unopened account.
func (a *Account) Withdraw(amt int64) (newBal int64, ok bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.opened || amt <= 0 {
		return 0, false
	}
	if a.balance < amt {
		return 0, false
	}
	a.balance -= amt
	return a.balance, true
}

// Close the account if opened.
// @returns: Payout of existing amount, and success flag
// @raises: Sucess flag false if not opened
func (a *Account) Close() (pay int64, ok bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.opened {
		return 0, false
	}
	a.opened = false
	payOut := a.balance
	a.balance = 0
	return payOut, true
}
